File: syscall/decode.c
/*
 * syscall/decode.c — Common syscall argument decoding helpers.
 * Arch-independent formatters for strings, flags, and per-syscall output.
 */
#define _GNU_SOURCE
#include "decode.h"
#include <string.h>
#include <sys/uio.h>
#include <fcntl.h>
#include <sys/mman.h>
#include <sys/socket.h>

/* ─── String reader ───────────────────────────────────────────────────────── */

int vock_read_str(pid_t pid, unsigned long addr, char *buf, size_t size)
{
	if (!addr || addr > 0x0000ffffffffffff) {
		buf[0] = '\0';
		return -1;
	}
	struct iovec local = { .iov_base = buf, .iov_len = size };
	struct iovec remote = { .iov_base = (void *)addr, .iov_len = size };
	ssize_t n = process_vm_readv(pid, &local, 1, &remote, 1, 0);
	if (n <= 0) { buf[0] = '\0'; return -1; }
	buf[size - 1] = '\0';
	return 0;
}

void vock_print_str(FILE *out, pid_t pid, unsigned long addr)
{
	char buf[256];
	if (vock_read_str(pid, addr, buf, sizeof(buf)) == 0 && buf[0]) {
		fputc('"', out);
		for (int i = 0; buf[i] && i < 64; i++) {
			if (buf[i] >= 32 && buf[i] < 127)
				fputc(buf[i], out);
			else
				fprintf(out, "\\x%02x", (unsigned char)buf[i]);
		}
		fputc('"', out);
	} else {
		if (addr == 0) fprintf(out, "NULL");
		else fprintf(out, "0x%lx", addr);
	}
}

/* ─── Flag printers ───────────────────────────────────────────────────────── */

void vock_print_open_flags(FILE *out, long flags)
{
	int mode = flags & 3;
	const char *m[] = {"O_RDONLY", "O_WRONLY", "O_RDWR", "O_RDWR"};
	fprintf(out, "%s", m[mode]);
	flags &= ~3;
	if (flags & O_CREAT)    { fprintf(out, "|O_CREAT"); flags &= ~O_CREAT; }
	if (flags & O_EXCL)     { fprintf(out, "|O_EXCL"); flags &= ~O_EXCL; }
	if (flags & O_TRUNC)    { fprintf(out, "|O_TRUNC"); flags &= ~O_TRUNC; }
	if (flags & O_APPEND)   { fprintf(out, "|O_APPEND"); flags &= ~O_APPEND; }
	if (flags & O_NONBLOCK) { fprintf(out, "|O_NONBLOCK"); flags &= ~O_NONBLOCK; }
	if (flags & O_CLOEXEC)  { fprintf(out, "|O_CLOEXEC"); flags &= ~O_CLOEXEC; }
	if (flags & O_DIRECTORY){ fprintf(out, "|O_DIRECTORY"); flags &= ~O_DIRECTORY; }
#ifdef O_LARGEFILE
	if (flags & O_LARGEFILE){ fprintf(out, "|O_LARGEFILE"); flags &= ~O_LARGEFILE; }
#endif
	if (flags) fprintf(out, "|0x%lx", flags);
}

void vock_print_mmap_prot(FILE *out, long prot)
{
	if (prot == 0) { fprintf(out, "PROT_NONE"); return; }
	int f = 1;
	if (prot & 1) { fprintf(out, "PROT_READ"); f = 0; prot &= ~1; }
	if (prot & 2) { fprintf(out, "%sPROT_WRITE", f ? "" : "|"); f = 0; prot &= ~2; }
	if (prot & 4) { fprintf(out, "%sPROT_EXEC", f ? "" : "|"); prot &= ~4; }
	if (prot) fprintf(out, "|0x%lx", prot);
}

void vock_print_mmap_flags(FILE *out, long flags)
{
	if (flags & MAP_PRIVATE) fprintf(out, "MAP_PRIVATE");
	else if (flags & MAP_SHARED) fprintf(out, "MAP_SHARED");
	else fprintf(out, "0x%lx", flags & 0xf);
	flags &= ~0xf;
	if (flags & MAP_ANONYMOUS) { fprintf(out, "|MAP_ANONYMOUS"); flags &= ~MAP_ANONYMOUS; }
	if (flags & MAP_FIXED)     { fprintf(out, "|MAP_FIXED"); flags &= ~MAP_FIXED; }
#ifdef MAP_POPULATE
	if (flags & MAP_POPULATE)  { fprintf(out, "|MAP_POPULATE"); flags &= ~MAP_POPULATE; }
#endif
	if (flags) fprintf(out, "|0x%lx", flags);
}

void vock_print_socket_domain(FILE *out, long d)
{
	switch (d) {
	case AF_UNIX: fprintf(out, "AF_UNIX"); break;
	case AF_INET: fprintf(out, "AF_INET"); break;
	case AF_INET6: fprintf(out, "AF_INET6"); break;
	case AF_NETLINK: fprintf(out, "AF_NETLINK"); break;
	case AF_PACKET: fprintf(out, "AF_PACKET"); break;
	default: fprintf(out, "%ld", d);
	}
}

void vock_print_socket_type(FILE *out, long type)
{
	long base = type & 0xf;
	switch (base) {
	case SOCK_STREAM: fprintf(out, "SOCK_STREAM"); break;
	case SOCK_DGRAM: fprintf(out, "SOCK_DGRAM"); break;
	case SOCK_RAW: fprintf(out, "SOCK_RAW"); break;
	case SOCK_SEQPACKET: fprintf(out, "SOCK_SEQPACKET"); break;
	default: fprintf(out, "%ld", base);
	}
	if (type & SOCK_NONBLOCK) fprintf(out, "|SOCK_NONBLOCK");
	if (type & SOCK_CLOEXEC) fprintf(out, "|SOCK_CLOEXEC");
}

/* ─── Per-syscall formatters ──────────────────────────────────────────────── */

void vock_fmt_openat(FILE *out, pid_t pid, long *a, long ret)
{
	if (a[0] == -100 || (unsigned long)a[0] == 0xffffff9c ||
	    (unsigned long)a[0] == 0xffffffffffffff9c)
		fprintf(out, "openat(AT_FDCWD, ");
	else
		fprintf(out, "openat(%ld, ", a[0]);
	vock_print_str(out, pid, a[1]);
	fprintf(out, ", ");
	vock_print_open_flags(out, a[2]);
	if (a[2] & O_CREAT) fprintf(out, ", %04lo", a[3]);
	fprintf(out, ") = %ld\n", ret);
}

void vock_fmt_open(FILE *out, pid_t pid, long *a, long ret)
{
	fprintf(out, "open(");
	vock_print_str(out, pid, a[0]);
	fprintf(out, ", ");
	vock_print_open_flags(out, a[1]);
	if (a[1] & O_CREAT) fprintf(out, ", %04lo", a[2]);
	fprintf(out, ") = %ld\n", ret);
}

void vock_fmt_read(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "read(%ld, 0x%lx, %lu) = %ld\n", a[0], a[1], a[2], ret);
}

void vock_fmt_write(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "write(%ld, 0x%lx, %lu) = %ld\n", a[0], a[1], a[2], ret);
}

void vock_fmt_close(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "close(%ld) = %ld\n", a[0], ret);
}

void vock_fmt_mmap(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "mmap(");
	if (a[0] == 0) fprintf(out, "NULL"); else fprintf(out, "0x%lx", a[0]);
	fprintf(out, ", %lu, ", a[1]);
	vock_print_mmap_prot(out, a[2]);
	fprintf(out, ", ");
	vock_print_mmap_flags(out, a[3]);
	fprintf(out, ", %ld, %ld) = 0x%lx\n", a[4], a[5], (unsigned long)ret);
}

void vock_fmt_mprotect(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "mprotect(0x%lx, %lu, ", a[0], a[1]);
	vock_print_mmap_prot(out, a[2]);
	fprintf(out, ") = %ld\n", ret);
}

void vock_fmt_socket(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "socket(");
	vock_print_socket_domain(out, a[0]);
	fprintf(out, ", ");
	vock_print_socket_type(out, a[1]);
	fprintf(out, ", %ld) = %ld\n", a[2], ret);
}

void vock_fmt_connect(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	fprintf(out, "connect(%ld, 0x%lx, %lu) = %ld\n", a[0], a[1], a[2], ret);
}

void vock_fmt_execve(FILE *out, pid_t pid, long *a, long ret)
{
	fprintf(out, "execve(");
	vock_print_str(out, pid, a[0]);
	fprintf(out, ", 0x%lx, 0x%lx) = %ld\n", a[1], a[2], ret);
}

void vock_fmt_access(FILE *out, pid_t pid, long *a, long ret)
{
	fprintf(out, "access(");
	vock_print_str(out, pid, a[0]);
	fprintf(out, ", ");
	long m = a[1];
	if (m == 0) fprintf(out, "F_OK");
	else {
		int f = 1;
		if (m & 4) { fprintf(out, "R_OK"); f = 0; m &= ~4; }
		if (m & 2) { fprintf(out, "%sW_OK", f ? "" : "|"); f = 0; m &= ~2; }
		if (m & 1) { fprintf(out, "%sX_OK", f ? "" : "|"); m &= ~1; }
	}
	fprintf(out, ") = %ld\n", ret);
}

void vock_fmt_brk(FILE *out, pid_t pid, long *a, long ret)
{
	(void)pid;
	if (a[0] == 0) fprintf(out, "brk(NULL) = 0x%lx\n", (unsigned long)ret);
	else fprintf(out, "brk(0x%lx) = 0x%lx\n", a[0], (unsigned long)ret);
}

void vock_fmt_generic(FILE *out, long nr, long *args, long ret)
{
	const char *name = vock_syscall_name(nr);
	if (name) fprintf(out, "%s(", name);
	else fprintf(out, "syscall_%ld(", nr);
	for (int i = 0; i < 6; i++) {
		if (i) fprintf(out, ", ");
		if (args[i] == 0) fprintf(out, "0");
		else fprintf(out, "0x%lx", (unsigned long)args[i]);
	}
	fprintf(out, ") = %ld\n", ret);
}

File: vock/src/syscall/decode.rs
//! strace-format syscall decoding (port of syscall/decode.c + arch decode.c).
//!
//! Reads strings from the traced process via `process_vm_readv` and renders
//! `name(args...) = ret` lines byte-for-byte compatible with the original.

use super::{read_str, syscall_name, Syscall};
use std::io::Write;

fn print_str<W: Write>(out: &mut W, pid: libc::pid_t, addr: u64) {
    match read_str(pid, addr, 256) {
        Some(buf) if !buf.is_empty() => {
            let _ = write!(out, "\"");
            for &c in buf.iter().take(64) {
                if (32..127).contains(&c) {
                    let _ = out.write_all(&[c]);
                } else {
                    let _ = write!(out, "\\x{c:02x}");
                }
            }
            let _ = write!(out, "\"");
        }
        _ => {
            if addr == 0 {
                let _ = write!(out, "NULL");
            } else {
                let _ = write!(out, "0x{addr:x}");
            }
        }
    }
}

fn print_open_flags<W: Write>(out: &mut W, flags: i64) {
    let mut flags = flags;
    let mode = (flags & 3) as usize;
    let m = ["O_RDONLY", "O_WRONLY", "O_RDWR", "O_RDWR"];
    let _ = write!(out, "{}", m[mode]);
    flags &= !3;
    for (bit, name) in [
        (libc::O_CREAT, "O_CREAT"),
        (libc::O_EXCL, "O_EXCL"),
        (libc::O_TRUNC, "O_TRUNC"),
        (libc::O_APPEND, "O_APPEND"),
        (libc::O_NONBLOCK, "O_NONBLOCK"),
        (libc::O_CLOEXEC, "O_CLOEXEC"),
        (libc::O_DIRECTORY, "O_DIRECTORY"),
        (libc::O_LARGEFILE, "O_LARGEFILE"),
    ] {
        let bit = bit as i64;
        if flags & bit != 0 {
            let _ = write!(out, "|{name}");
            flags &= !bit;
        }
    }
    if flags != 0 {
        let _ = write!(out, "|0x{flags:x}");
    }
}

fn print_mmap_prot<W: Write>(out: &mut W, prot: i64) {
    let mut prot = prot;
    if prot == 0 {
        let _ = write!(out, "PROT_NONE");
        return;
    }
    let mut first = true;
    for (bit, name) in [(1, "PROT_READ"), (2, "PROT_WRITE"), (4, "PROT_EXEC")] {
        if prot & bit != 0 {
            let _ = write!(out, "{}{}", if first { "" } else { "|" }, name);
            first = false;
            prot &= !bit;
        }
    }
    if prot != 0 {
        let _ = write!(out, "|0x{prot:x}");
    }
}

fn print_mmap_flags<W: Write>(out: &mut W, flags: i64) {
    let mut flags = flags;
    if flags & libc::MAP_PRIVATE as i64 != 0 {
        let _ = write!(out, "MAP_PRIVATE");
    } else if flags & libc::MAP_SHARED as i64 != 0 {
        let _ = write!(out, "MAP_SHARED");
    } else {
        let _ = write!(out, "0x{:x}", flags & 0xf);
    }
    flags &= !0xf;
    for (bit, name) in [
        (libc::MAP_ANONYMOUS as i64, "MAP_ANONYMOUS"),
        (libc::MAP_FIXED as i64, "MAP_FIXED"),
        (libc::MAP_POPULATE as i64, "MAP_POPULATE"),
    ] {
        if flags & bit != 0 {
            let _ = write!(out, "|{name}");
            flags &= !bit;
        }
    }
    if flags != 0 {
        let _ = write!(out, "|0x{flags:x}");
    }
}

fn print_socket_domain<W: Write>(out: &mut W, d: i64) {
    let s = match d {
        x if x == libc::AF_UNIX as i64 => "AF_UNIX",
        x if x == libc::AF_INET as i64 => "AF_INET",
        x if x == libc::AF_INET6 as i64 => "AF_INET6",
        x if x == libc::AF_NETLINK as i64 => "AF_NETLINK",
        x if x == libc::AF_PACKET as i64 => "AF_PACKET",
        _ => {
            let _ = write!(out, "{d}");
            return;
        }
    };
    let _ = write!(out, "{s}");
}

fn print_socket_type<W: Write>(out: &mut W, ty: i64) {
    let base = ty & 0xf;
    let s = match base {
        x if x == libc::SOCK_STREAM as i64 => "SOCK_STREAM",
        x if x == libc::SOCK_DGRAM as i64 => "SOCK_DGRAM",
        x if x == libc::SOCK_RAW as i64 => "SOCK_RAW",
        x if x == libc::SOCK_SEQPACKET as i64 => "SOCK_SEQPACKET",
        _ => "",
    };
    if s.is_empty() {
        let _ = write!(out, "{base}");
    } else {
        let _ = write!(out, "{s}");
    }
    if ty & libc::SOCK_NONBLOCK as i64 != 0 {
        let _ = write!(out, "|SOCK_NONBLOCK");
    }
    if ty & libc::SOCK_CLOEXEC as i64 != 0 {
        let _ = write!(out, "|SOCK_CLOEXEC");
    }
}

// ─── Per-syscall formatters ─────────────────────────────────────────────────

fn is_at_fdcwd(v: i64) -> bool {
    v == -100 || (v as u64) == 0xffff_ff9c || (v as u64) == 0xffff_ffff_ffff_ff9c
}

fn fmt_openat<W: Write>(out: &mut W, pid: libc::pid_t, a: &[i64; 6], ret: i64) {
    if is_at_fdcwd(a[0]) {
        let _ = write!(out, "openat(AT_FDCWD, ");
    } else {
        let _ = write!(out, "openat({}, ", a[0]);
    }
    print_str(out, pid, a[1] as u64);
    let _ = write!(out, ", ");
    print_open_flags(out, a[2]);
    if a[2] & libc::O_CREAT as i64 != 0 {
        let _ = write!(out, ", {:04o}", a[3]);
    }
    let _ = writeln!(out, ") = {ret}");
}

fn fmt_open<W: Write>(out: &mut W, pid: libc::pid_t, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "open(");
    print_str(out, pid, a[0] as u64);
    let _ = write!(out, ", ");
    print_open_flags(out, a[1]);
    if a[1] & libc::O_CREAT as i64 != 0 {
        let _ = write!(out, ", {:04o}", a[2]);
    }
    let _ = writeln!(out, ") = {ret}");
}

fn fmt_read<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = writeln!(out, "read({}, 0x{:x}, {}) = {}", a[0], a[1] as u64, a[2] as u64, ret);
}

fn fmt_write<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = writeln!(out, "write({}, 0x{:x}, {}) = {}", a[0], a[1] as u64, a[2] as u64, ret);
}

fn fmt_close<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = writeln!(out, "close({}) = {}", a[0], ret);
}

fn fmt_mmap<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "mmap(");
    if a[0] == 0 {
        let _ = write!(out, "NULL");
    } else {
        let _ = write!(out, "0x{:x}", a[0] as u64);
    }
    let _ = write!(out, ", {}, ", a[1] as u64);
    print_mmap_prot(out, a[2]);
    let _ = write!(out, ", ");
    print_mmap_flags(out, a[3]);
    let _ = writeln!(out, ", {}, {}) = 0x{:x}", a[4], a[5], ret as u64);
}

fn fmt_mprotect<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "mprotect(0x{:x}, {}, ", a[0] as u64, a[1] as u64);
    print_mmap_prot(out, a[2]);
    let _ = writeln!(out, ") = {ret}");
}

fn fmt_socket<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "socket(");
    print_socket_domain(out, a[0]);
    let _ = write!(out, ", ");
    print_socket_type(out, a[1]);
    let _ = writeln!(out, ", {}) = {}", a[2], ret);
}

fn fmt_connect<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    let _ = writeln!(out, "connect({}, 0x{:x}, {}) = {}", a[0], a[1] as u64, a[2] as u64, ret);
}

fn fmt_execve<W: Write>(out: &mut W, pid: libc::pid_t, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "execve(");
    print_str(out, pid, a[0] as u64);
    let _ = writeln!(out, ", 0x{:x}, 0x{:x}) = {}", a[1] as u64, a[2] as u64, ret);
}

fn fmt_access<W: Write>(out: &mut W, pid: libc::pid_t, a: &[i64; 6], ret: i64) {
    let _ = write!(out, "access(");
    print_str(out, pid, a[0] as u64);
    let _ = write!(out, ", ");
    let mut m = a[1];
    if m == 0 {
        let _ = write!(out, "F_OK");
    } else {
        let mut first = true;
        for (bit, name) in [(4, "R_OK"), (2, "W_OK"), (1, "X_OK")] {
            if m & bit != 0 {
                let _ = write!(out, "{}{}", if first { "" } else { "|" }, name);
                first = false;
                m &= !bit;
            }
        }
    }
    let _ = writeln!(out, ") = {ret}");
}

fn fmt_brk<W: Write>(out: &mut W, a: &[i64; 6], ret: i64) {
    if a[0] == 0 {
        let _ = writeln!(out, "brk(NULL) = 0x{:x}", ret as u64);
    } else {
        let _ = writeln!(out, "brk(0x{:x}) = 0x{:x}", a[0] as u64, ret as u64);
    }
}

fn fmt_generic<W: Write>(out: &mut W, nr: i64, args: &[i64; 6], ret: i64) {
    match syscall_name(nr) {
        Some(name) => {
            let _ = write!(out, "{name}(");
        }
        None => {
            let _ = write!(out, "syscall_{nr}(");
        }
    }
    for (i, &arg) in args.iter().enumerate() {
        if i != 0 {
            let _ = write!(out, ", ");
        }
        if arg == 0 {
            let _ = write!(out, "0");
        } else {
            let _ = write!(out, "0x{:x}", arg as u64);
        }
    }
    let _ = writeln!(out, ") = {ret}");
}

// ─── Arch dispatch (x86_64 has legacy syscalls absent on aarch64) ───────────

#[cfg(target_arch = "x86_64")]
mod nr {
    pub const READ: i64 = 0;
    pub const WRITE: i64 = 1;
    pub const OPEN: i64 = 2;
    pub const CLOSE: i64 = 3;
    pub const MMAP: i64 = 9;
    pub const MPROTECT: i64 = 10;
    pub const BRK: i64 = 12;
    pub const ACCESS: i64 = 21;
    pub const SOCKET: i64 = 41;
    pub const CONNECT: i64 = 42;
    pub const EXECVE: i64 = 59;
    pub const OPENAT: i64 = 257;
}

#[cfg(target_arch = "aarch64")]
mod nr {
    pub const OPENAT: i64 = 56;
    pub const CLOSE: i64 = 57;
    pub const READ: i64 = 63;
    pub const WRITE: i64 = 64;
    pub const BRK: i64 = 214;
    pub const MMAP: i64 = 222;
    pub const MPROTECT: i64 = 226;
    pub const EXECVE: i64 = 221;
    pub const SOCKET: i64 = 198;
    pub const CONNECT: i64 = 203;
}

/// Decode one syscall into strace format, writing a single line to `out`.
#[cfg(target_arch = "x86_64")]
pub fn decode_syscall<W: Write>(out: &mut W, pid: libc::pid_t, sc: &Syscall) {
    let a = &sc.args;
    match sc.nr {
        nr::READ => fmt_read(out, a, sc.ret),
        nr::WRITE => fmt_write(out, a, sc.ret),
        nr::OPEN => fmt_open(out, pid, a, sc.ret),
        nr::CLOSE => fmt_close(out, a, sc.ret),
        nr::MMAP => fmt_mmap(out, a, sc.ret),
        nr::MPROTECT => fmt_mprotect(out, a, sc.ret),
        nr::BRK => fmt_brk(out, a, sc.ret),
        nr::ACCESS => fmt_access(out, pid, a, sc.ret),
        nr::SOCKET => fmt_socket(out, a, sc.ret),
        nr::CONNECT => fmt_connect(out, a, sc.ret),
        nr::EXECVE => fmt_execve(out, pid, a, sc.ret),
        nr::OPENAT => fmt_openat(out, pid, a, sc.ret),
        _ => fmt_generic(out, sc.nr, a, sc.ret),
    }
}

#[cfg(target_arch = "aarch64")]
pub fn decode_syscall<W: Write>(out: &mut W, pid: libc::pid_t, sc: &Syscall) {
    let a = &sc.args;
    match sc.nr {
        nr::READ => fmt_read(out, a, sc.ret),
        nr::WRITE => fmt_write(out, a, sc.ret),
        nr::OPENAT => fmt_openat(out, pid, a, sc.ret),
        nr::CLOSE => fmt_close(out, a, sc.ret),
        nr::MMAP => fmt_mmap(out, a, sc.ret),
        nr::MPROTECT => fmt_mprotect(out, a, sc.ret),
        nr::BRK => fmt_brk(out, a, sc.ret),
        nr::SOCKET => fmt_socket(out, a, sc.ret),
        nr::CONNECT => fmt_connect(out, a, sc.ret),
        nr::EXECVE => fmt_execve(out, pid, a, sc.ret),
        _ => fmt_generic(out, sc.nr, a, sc.ret),
    }
}
